### PR TITLE
Feature resource mover script

### LINKED ISSUE #1788: Create script to move string resources across modules
**Is your feature request related to a problem? Please describe.**
To support our translations we lint  the core module as a dependent of the app module. This has a side effect of linting all of the dependencies of app and creates false positives in our lint reports that have to be baselined

**Describe the solution you'd like**
A script that takes 3 inputs, `source module`, `destination module` `string key` and uses them to move a specific string and all its translation to the `destination module` so our translations are preserved on translatewiki

**Describe alternatives you've considered**
Continue with how it is, downside is we will never get rid of the baseline

### PR DESCRIPTION
**Pull request by Frans-Lukas**
Fixes #1788

Rewrote move-string-resource script and added executable binary of script. 

instrumentation.sh, testdroid.py, update-play-store.py and upload-apk.py were moved to contrib folder from root.

### CODE REVIEW

**Comment by macgills** [comment]
The instrumentation tests are failing because `instrumentation.sh` was moved
update `coverage.yml` so it invokes `contrib/instrumentation.sh`

File: contrib/move-string-resource.kts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+if (args.size < 3) {
+  printCorrectUsageAndExit()
+}
+
+val pathToValues = "/src/main/res/"
+val source = args[0]
+val destination = args[1]
+
+/// name=key
```
Comment (macgills) [inline contrib/move-string-resource.kts]: delete this

File: contrib/move-string-resource.kts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+println("Running transfer of string resources...")
+val numberOfFilesPastedTo = sourceDir.cutStringResourcesAndPasteToDestination(key, source, destination)
+println("Transfer of string resources complete. Moved $numberOfFilesPastedTo strings.")
+
+fun File.cutStringResourcesAndPasteToDestination(key: String, source: String, destination: String) : Int {
+  var numberOfFilesPastedTo = 0
+  this.walk().filter { it.name.equals("strings.xml") }.forEach { resourceFile ->
+    var cutLine: String = cutLineFromResourceFile(resourceFile, key)
```
Comment (macgills) [inline contrib/move-string-resource.kts]: lets make this a little more kotlin-y
```
cutLineFromResourceFile(resourceFile, key).takeif{ it.isNotEmpty() }?.let{ cutLine ->
  pasteLineToDestination(
    openOrCreateDestinationResourceFile(resourceFile, source, destination), 
    cutLine
  )
}
```

File: contrib/move-string-resource.kts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    val destinationFile = openOrCreateDestinationResourceFile(resourceFile, source, destination)
+    pasteLineToDestination(destinationFile, cutLine)
+    numberOfFilesPastedTo++
+  }
+  return numberOfFilesPastedTo
+}
+
+fun openOrCreateDestinationResourceFile(resourceFile: File, source: String, destination: String): File {
+
+  val destinationFilePath = resourceFile.path.replaceRange(0..source.length, destination + "/")
```
Comment (macgills) [inline contrib/move-string-resource.kts]: Was `mkdirs' giving you trouble?
image

File: contrib/move-string-resource.kts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  val destinationFile = File(destinationFilePath)
+  val destinationDirectoryPath = destinationFilePath
```
Comment (macgills) [inline contrib/move-string-resource.kts]: File has a property `parentFile` if you want to get the file above it

File: contrib/move-string-resource.kts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+fun createDestinationDirectoryAndFile(destinationDirectory: File, destinationFile: File) {
+  destinationDirectory.mkdir()
+  val isNewFileCreated: Boolean = destinationFile.createNewFile()
+  if (!isNewFileCreated) {
+    System.err.println("Could not create resource file ${destinationFile.path}")
+    System.exit(-1)
+  }
+  val initialXMLResourceFileContent = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
```
Comment (macgills) [inline contrib/move-string-resource.kts]: here we can definitely use a raw string essentailly what they are made for.
"""
<?xml version=\"1.0\" encoding=\"UTF-8\"?>
<resources>
</resources>
""".trimIdent()

File: contrib/move-string-resource.kts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    "<resources xmlns:tools=\"http://schemas.android.com/tools\" tools:ignore=\"DuplicateStrings\">\n" +
```
Comment (macgills) [inline contrib/move-string-resource.kts]: we don't want to by default ignore DuplicateStrings

File: contrib/move-string-resource.kts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    "</resources>"
+  destinationFile.writeText(initialXMLResourceFileContent)
+  println("Created directory ${destinationDirectory.path} and resource file ${destinationFile.path}")
+}
+
+fun pasteLineToDestination(destinationFile: File, cutLine: String) {
+  var resourceDataWithPastedLine = ""
```
Comment (macgills) [inline contrib/move-string-resource.kts]: So using a StringBuilder here will increase readability I think

File: contrib/move-string-resource.kts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+  var firstLineRead = false
```
Comment (macgills) [inline contrib/move-string-resource.kts]: I don't think you need this.
to me the flow is 
```
read line
add line to stringbuilder
if (line was "<resources>") add cutLine
```
I might be reading it wrong but it looks like `Add new line` just starts the file off with a `/n`

Comment (Frans-Lukas) [reply contrib/move-string-resource.kts]: I don't think `read line` keeps newlines, my solution was the firstLineRead bool and to not put newlines if it is false. Couldn't think of a cleaner solution, but if your suggestion works (read line keeps line breaks) I think it is better. 

``` 
fun addNewLine(firstLineRead: Boolean, stringToAddLineTo: String, line: String): String {
...
  if (firstLineRead) {
    tempString = tempString + "\n"
  }
...
```

**Comment by Frans-Lukas** [comment]
@macgills Again, great feedback. Thanks!

File: contrib/move-string-resource.kts (hunk 8 above)
Comment (macgills) [reply contrib/move-string-resource.kts]: StringBuilder has `appendLn` functions, I think that is what you'll want. My memory is failing me if readLine preserves line endings

File: contrib/move-string-resource.kts (hunk 5 above)
Comment (macgills) [reply contrib/move-string-resource.kts]: I hadn't formatted this properly so it was hiding the contents of the raw string, edited

File: contrib/move-string-resource.kts (hunk 8 above)
Comment (Frans-Lukas) [reply contrib/move-string-resource.kts]: ok nice! That will look better.

File: contrib/move-string-resource.kts [hunk 9]
```diff
@@ -86,49 +83,48 @@ fun openOrCreateDestinationResourceFile(resourceFile: File, source: String, dest
...
+  destinationDirectory.mkdirs()
   val isNewFileCreated: Boolean = destinationFile.createNewFile()
   if (!isNewFileCreated) {
     System.err.println("Could not create resource file ${destinationFile.path}")
     System.exit(-1)
   }
-  val initialXMLResourceFileContent = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
-    "<resources xmlns:tools=\"http://schemas.android.com/tools\" tools:ignore=\"DuplicateStrings\">\n" +
-    "</resources>"
+  val initialXMLResourceFileContent = """
```
Comment (macgills) [inline contrib/move-string-resource.kts]: I usually tab these so the ide shows a line running between the `"""` pairs, so I generally start them on a new line after the `=`. Having yet to write a kotlin script myself I am not sure if the highlighting is quite up to scratch but I imagine it should be

I would also be in favour of this being either not being a variable or be declared as a val at the top of the file.
I think inline is fine really
```
 destinationFile.writeText(
"""
    <?xml version=\"1.0\" encoding=\"UTF-8\"?>
    <resources>
    </resources>
""".trimIdent()
)
```
is pretty self explanatory to me.

**Comment by macgills** [comment]
Good work @Frans-Lukas , it is nearly there. Thank you for tolerating my nitpicking. Just one more change, a recompile of the standalone binary and we should all be done.

**Comment by Frans-Lukas** [comment]
> 
> 
> Good work @Frans-Lukas , it is nearly there. Thank you for tolerating my nitpicking. Just one more change, a recompile of the standalone binary and we should all be done.

No worries, it is very helpful to get your feedback. I am somewhat new to kotlin so I am getting a lot out of it. It's also fun to write cleaner code.

**Comment by macgills** [comment]
Well if you want to practice kotlin further 41.8% of the project is in Java and I'd love to see that number hit 0.
I'll be making a ticket soon inviting contributors to create a ticket to convert a specific java class (I recommend small ones to start with) to kotlin, there will also be a similar ticket with respect to unit testing

**Comment by Frans-Lukas** [comment]
Cool! That also sounds like a good way to learn the structure of the project.

File: contrib/move-string-resource.kts (hunk 9 above)
Comment (macgills) [reply contrib/move-string-resource.kts]: @Frans-Lukas this is my only remaining comment

Comment (macgills) [reply contrib/move-string-resource.kts]: @Frans-Lukas this comment is still outstanding

Comment (Frans-Lukas) [reply contrib/move-string-resource.kts]: oh! missed that. On it.

Comment (Frans-Lukas) [reply contrib/move-string-resource.kts]: @macgills Should be fixed now 👍

Comment (macgills) [reply contrib/move-string-resource.kts]: Excellent, we will get to merge this as soon as the build passes

**Pull request merged**